### PR TITLE
Refactored the way we retrieve students ethnicity.

### PR DESCRIPTION
**Pull request by agorina**
Why?  There are two ways we get students ethnicity: in one repository we used `GROUP_CONCAT` and in another we looked up the ethnicity as a second database call. I need a similar logic for the target report API and I wanted to re-use `GROUP_CONCAT` solution. When I tested it on AWS Dev, it was **very** slow ( I had to kill the query since it was not responding for a while). The main offender is `GROUP BY`. So I changed it to do a second call.

### CODE REVIEW

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/AbstractExamRepository.java [hunk 1]
```diff
@@ -94,25 +91,5 @@ protected AbstractExamRepository(
         return exams;
     }
 
-    private Map<Long, Set<String>> findEthnicityCodesForStudentIds(final Collection<Long> studentIds) {
```
Comment (agorina) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/AbstractExamRepository.java]: Moved it into a separate repository to be reused.

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/JdbcStudentRepository.java [hunk 2]
```diff
@@ -9,51 +9,71 @@
...
 
     @Override
     public List<Student> findAll(@NotNull final PermissionSource permissionSource,
                                  @NotNull final StudentSearch search) {
-        return template.query(
+
+        final List<Student.Builder> studentBuilders = newArrayList();
+        final Set<Long> studentIds = newHashSet();
+
+        // SQL returns a sorted list, need to preserve the order
```
Comment (agorina) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/JdbcStudentRepository.java]: @esotrope - thank you for writing a test for this!

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/Student.java [hunk 3]
```diff
@@ -128,6 +128,10 @@ public Builder copy() {
         private Boolean section504;
         private Boolean migrantStatus;
 
+        public long getId() {
+            return id;
+        }
```
Comment (agorina) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/Student.java]: I know, but..I need it.

File: reporting-service/src/test/java/org/opentestsystem/rdw/reporting/TestData.java [hunk 4]
```diff
@@ -153,7 +153,7 @@
                 .firstName("student1_firstName")
                 .lastName("student1_lastName")
                 .genderCode("g1")
-                .ethnicityCodes(ImmutableSet.of());
+                .ethnicityCodes(ImmutableSet.of("ethnicity-29", "ethnicity-28"));
```
Comment (agorina) [inline reporting-service/src/test/java/org/opentestsystem/rdw/reporting/TestData.java]: I added test data.

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/JdbcStudentRepository.java [hunk 5]
```diff
@@ -9,51 +9,71 @@
...
+                            .limitedEnglishProficiency(getNullable(row, row.getBoolean("lep")))
+                            .section504(getNullable(row, row.getBoolean("section504")))
+                            .migrantStatus(getNullable(row, row.getBoolean("migrant_status"))));
+                    return row;
+                });
 
+        final Map<Long, Set<String>> ethnicityCodesByStudentId = ethnicityRepository.findEthnicityCodesByStudentIds(studentIds);
+        return studentBuilders
+                .stream()
+                .map(b -> b.ethnicityCodes(ethnicityCodesByStudentId.getOrDefault(b.getId(), null)).build())
```
Comment (esotrope) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/JdbcStudentRepository.java]: doesn't Map.get() return null when the key is absent?

Comment (agorina) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/JdbcStudentRepository.java]: Good point, changing.

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/JdbcStudentEthnicityRepository.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                row -> {
+                    final long studentId = row.getLong("student_id");
+                    final String ethnicityCode = row.getString("ethnicity_code");
+
+                    final Set<String> studentEthnicityCodes = ethnicityCodesByStudentId.get(studentId);
+                    if (studentEthnicityCodes == null) {
+                        ethnicityCodesByStudentId.put(studentId, newHashSet(ethnicityCode));
+                    } else {
+                        studentEthnicityCodes.add(ethnicityCode);
+                    }
```
Comment (malaffoon) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/student/JdbcStudentEthnicityRepository.java]: nit: i think the computeIfAbsent approach is more readable, i.e.:
```
        final Set<String> studentEthnicityCodes = ethnicityCodesByStudentId.computeIfAbsent(studentId, newHashSet());
        studentEthnicityCodes.add(ethnicityCode);
```

**Pull request merged**